sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -1115,9 +1126,38 @@ object SQLContext {
             })
         }
     }
-    def createConverter(cls: Class[_], dataType: DataType): Any => Any = dataType match {
-      case struct: StructType => createStructConverter(cls, struct.map(_.dataType))
-      case _ => CatalystTypeConverters.createToCatalystConverter(dataType)
+    def createConverter(t: Type, dataType: DataType): Any => Any = (t, dataType) match {
+      case (cls: Class[_], struct: StructType) =>
+        // bean type
+        createStructConverter(cls, struct.map(_.dataType))
+      case (arrayType: Class[_], array: ArrayType) if arrayType.isArray =>
+        // array type
+        val converter = createConverter(arrayType.getComponentType, array.elementType)
+        value => new GenericArrayData(
+          (0 until JavaArray.getLength(value)).map(i =>
+            converter(JavaArray.get(value, i))).toArray)
+      case (_, array: ArrayType) =>
+        // java.util.List type
+        val cls = classOf[java.util.List[_]]
```

Comment:
Seems like `JavaTypeInference.inferDataType()` supports `java.lang.Iterable`, not only `List`, but serializer/deserializer don't. Should we change `inferDataType()`?

Reply:
I think you are right. It should be better to change it to avoid confusion. I also agree with a separate PR for that.